File: solutions/go/minesweeper/1/minesweeper.go
package minesweeper

// Annotate returns an annotated board
func Annotate(board []string) []string {
	if len(board) == 0 {
		return board
	}
	height := len(board)
	width := len(board[0])
	// if any(len(row) != width for row in board){
	//     return ValueError("The board is invalid with current input.")}

	annotate_cell := func(I, J int) byte {
		if board[I][J] != ' ' {
			return board[I][J]
		}
		var mines byte = 0
		for i := I - 1; i <= I+1; i++ {
			if !(0 <= i && i < height) {
				continue
			}
			for j := J - 1; j <= J+1; j++ {
				if !(0 <= j && j < width) ||
					(i == I && j == J) {
					continue
				}
				if board[i][j] == '*' {
					mines++
				}
			}
		}
		if mines == 0 {
			return ' '
		}
		return '0' + mines
	}

	for i := range board {
		row := []byte{}
		for j := range width {
			row = append(row, annotate_cell(i, j))
		}
		board[i] = string(row)
	}
	return board
}

File: solutions/haskell/minesweeper/2/src/Minesweeper.hs
module Minesweeper (annotate) where

annotate :: [String] -> [String]
annotate board =
    [ [ case cell of
            ' ' | let m = countMines r c, m > 0 -> head $ show m
            _ -> cell
      | (c, cell) <- zip [0 ..] row
      ]
    | (r, row) <- zip [0 ..] board
    ]
  where
    height = length board
    width = length $ head board
    countMines r c =
        length
            [ ()
            | i <- [max 0 (r - 1) .. min (height - 1) (r + 1)]
            , j <- [max 0 (c - 1) .. min (width - 1) (c + 1)]
            , (i, j) /= (r, c)
            , board !! i !! j == '*'
            ]
